modules/trycompile/src/Interrupt.cxx

```diff
@@ -0,0 +1,188 @@
+#   define SMCE_STDLIB_LIBCXX 1
+#   if defined(_LIBCPP_HAS_NO_THREADS)
+#       error "Detected LLVM libc++ without thread support"
+#   elif defined(_LIBCPP_HAS_THREAD_API_PTHREAD)
+#       define SMCE_THREADS_POSIX 1
+#   elif defined(_LIBCPP_HAS_THREAD_API_WIN32)
+#       define SMCE_THREADS_WIN32
+#   endif
+#elif defined(_YVALS) // Microsoft STL or Dinkumware
+#   define SMCE_STDLIB_MSSTL_OR_DINKUMWARE
+#   define SMCE_THREADS_WIN32 1
+#else
+#   error "Unknown/unsupported C++ standard library"
+#endif
+
+
+
+
+/* Platform-specific headers */
+#ifdef SMCE_THREADS_POSIX
+#if !__has_include(<pthread.h>)
+#error "PThreads chosen but header not available"
+#endif
+#include <pthread.h>
+#include <unistd.h>
+#endif
+#ifdef SMCE_THREADS_WIN32
+#include <Windows.h>
+#endif
+
+/* Implementations */
+#include <detail/Interrupt.hxx>
+
+namespace smce::detail {
+
+thread_local std::recursive_mutex* gtl_interrupt_mutex{};
+
+#ifdef SMCE_THREADS_POSIX
+#if defined(_POSIX_C_SOURCE) || defined(_XOPEN_SOURCE) || defined(__APPLE__)
+#if _POSIX_C_SOURCE >= 199506L || _XOPEN_SOURCE >= 500 || defined(__APPLE__)
+
+static_assert(std::is_same_v<std::thread::native_handle_type, pthread_t>);
+
+bool make_self_suspendable(std::recursive_mutex& guard) noexcept {
+    // For termination
+    int unused;
+    if(pthread_setcancelstate(PTHREAD_CANCEL_ENABLE, &unused) != 0)
+        return false;
+    if(pthread_setcanceltype(PTHREAD_CANCEL_ASYNCHRONOUS, &unused) != 0)
+        return false;
+
+    // For suspension
+    struct sigaction sa;
+    sa.sa_flags = 0;
+    sa.sa_handler = +[](int sig) noexcept -> void {
+        switch (sig) {
+        case SIGUSR1:
+            pause();
+            break;
+        case SIGUSR2:
+            break;
+        case SIGTERM:
+            pthread_exit(nullptr);
+        }
+    };
+
+    if (sigemptyset(&sa.sa_mask) != 0)
+        return false;
+    if (sigaction(SIGUSR1, &sa, nullptr) == -1)
+        return false;
+    if (sigaction(SIGUSR2, &sa, nullptr) == -1)
+        return false;
+    if (sigaction(SIGTERM, &sa, nullptr) == -1)
+        return false;
+
+    gtl_interrupt_mutex = &guard;
+    return true;
+}
+
+bool suspend_thread(std::thread& th) noexcept {
+    if (!gtl_interrupt_mutex)
+        return false;
+    gtl_interrupt_mutex->lock();
+    const auto res = pthread_kill(th.native_handle(), SIGUSR1);
+    if (res != 0)
+        gtl_interrupt_mutex->unlock();
+    return res == 0;
+}
+
+bool resume_thread(std::thread& th) noexcept {
+    if (!gtl_interrupt_mutex)
+        return false;
+    const auto res = pthread_kill(th.native_handle(), SIGUSR2);
+    if (res != 0)
+        return false;
+    gtl_interrupt_mutex->unlock();
+    return true;
+}
+
+bool murder_thread(std::thread& th) noexcept {
+    const pthread_t hdl = th.native_handle();
+    if(pthread_cancel(hdl) != 0)
+        return false;
+    void* res{};
+    if(pthread_join(hdl, &res) != 0)
+        return false;
+    return res == PTHREAD_CANCELED;
+}
```

Comment:
I haven't followed up on the rest of the project, but I am curious on why you're using `pthreads` instead of C++ threads. Is it for compatibility across OS?

Reply:
We are just hijacking the `std::thread` object because we need to suspend the threads and that is not standardized whatsoever (well, `pthread_suspend` exists, but nobody implements it, so WG21 didn't get it in C++0x)

Reply:
And while we're at it I figured we might as well use the same for thread termination because ISO C++ cannot terminate a single thread, instead just invoking `std::terminate` on the whole process (we don't want the main thread to die due to the user-code thread)

Reply:
Alright, I imagined something like that. Good to know! :+1: